torch/package/importer.py

```diff
@@ -0,0 +1,193 @@
+import importlib
+from abc import ABC, abstractmethod
+from pickle import _getattribute, whichmodule, _Pickler  # type: ignore
+from types import ModuleType
+from typing import Any, Optional, Tuple, List
+
+from ._mangling import demangle, get_mangle_prefix, is_mangled
+
+
+class ObjNotFoundError(Exception):
+    """Raised when an importer cannot find an object by searching for its name."""
+
+    pass
+
+
+class ObjMismatchError(Exception):
+    """Raised when an importer found a different object with the same name as the user-provided one."""
+
+    pass
+
+
+class Importer(ABC):
+    """Represents an environment to import modules from.
+
+    By default, you can figure out what module an object belongs by checking
+    __module__ and importing the result using __import__ or importlib.import_module.
+
+    torch.package introduces module importers other than the default one.
+    Each PackageImporter introduces a new namespace. Potentially a single
+    name (e.g. 'foo.bar') is present in multiple namespaces.
+
+    It supports two main operations:
+        import_module: module_name -> module object
+        get_name: object -> (parent module name, name of obj within module)
+
+    The guarantee is that following round-trip will succeed or throw an ObjNotFoundError/ObjMisMatchError.
+        module_name, obj_name = env.get_name(obj)
+        module = env.import_module(module_name)
+        obj2 = getattr(module, obj_name)
+        assert obj1 is obj2
+    """
+
+    @abstractmethod
+    def import_module(self, module_name: str) -> ModuleType:
+        """Import `module_name` from this environment.
+
+        The contract is the same as for importlib.import_module.
+        """
+        pass
+
+    def get_name(self, obj: Any, name: Optional[str] = None) -> Tuple[str, str]:
+        """Given an object, return a name that can be used to retrieve the
+        object from this environment.
+
+        Args:
+            obj: An object to get the the module-environment-relative name for.
+            name: If set, use this name instead of looking up __name__ or __qualname__ on `obj`.
+                This is only here to match how Pickler handles __reduce__ functions that return a string,
+                don't use otherwise.
+        Returns:
+            A tuple (parent_module_name, attr_name) that can be used to retrieve `obj` from this environment.
+            Use it like:
+                mod = importer.import_module(parent_module_name)
+                obj = getattr(mod, attr_name)
+
+        Raises:
+            ObjNotFoundError: we couldn't retrieve `obj by name.
+            ObjMisMatchError: we found a different object with the same name as `obj`.
+        """
+        if name is None and obj and _Pickler.dispatch.get(type(obj)) is None:
```

Comment:
Because this is the start of this function, and because it is only called in one place from our custom pickler, I think it  should go in that function and not here. I can see this accidentally firing for weird objects in other places.

Reply:
Does that make sense or am I not understanding what the string version of reduce does?

Reply:
afaict we need this here, because `__reduce__ -> str` is one way that objects can signal that they want to be global'd as a different name than what the pickler might infer. I ran across this when trying to package references to `typing.List` and friends in FX.

We can put equivalent logic in the import string formatter in `GraphModule`, but it made more sense to put it here and leave clients of the code unaware of this detail.

Reply:
ok, I get it now. Given that definition of what it does, then it is in the right place.